### PR TITLE
Add function for a normal dotted into a tensor

### PR DESCRIPTION
**Pull request by nilsvu**
### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/DataStructures/Tensor/EagerMath/DotProduct.hpp [hunk 1]
```diff
@@ -75,6 +75,50 @@ Scalar<DataType> dot_product(
...
+             tensor_right.get(prepend(dot_product_indices, d));
+    }
+  }
+}
+
+template <typename DataType, typename Index, typename Symm,
+          typename... RemainingIndices,
+          typename ResultTensor = Tensor<DataType, tmpl::pop_front<Symm>,
+                                         index_list<RemainingIndices...>>>
+ResultTensor dot_product(
```
Comment (kidder) [inline src/DataStructures/Tensor/EagerMath/DotProduct.hpp]: remove this function, we only want the not_null version

File: src/DataStructures/Tensor/EagerMath/DotProduct.hpp [hunk 2]
```diff
@@ -75,6 +75,50 @@ Scalar<DataType> dot_product(
...
+          typename... RemainingIndices,
+          typename ResultTensor = Tensor<DataType, tmpl::pop_front<Symm>,
+                                         index_list<RemainingIndices...>>>
+void dot_product(
+    const gsl::not_null<ResultTensor*> dot_product,
+    const Tensor<DataType, Symmetry<1>, index_list<Index>>& vector_left,
+    const Tensor<DataType, Symm,
+                 index_list<change_index_up_lo<Index>, RemainingIndices...>>&
+        tensor_right) noexcept {
+  *dot_product = make_with_value<ResultTensor>(vector_left, 0.);
```
Comment (kidder) [inline src/DataStructures/Tensor/EagerMath/DotProduct.hpp]: remove this

File: src/DataStructures/Tensor/EagerMath/DotProduct.hpp [hunk 3]
```diff
@@ -75,6 +75,50 @@ Scalar<DataType> dot_product(
...
+  for (auto it = dot_product->begin(); it != dot_product->end(); ++it) {
+    const auto dot_product_indices = dot_product->get_tensor_index(it);
+    for (size_t d = 0; d < Index::dim; d++) {
```
Comment (kidder) [inline src/DataStructures/Tensor/EagerMath/DotProduct.hpp]: unroll the first element as in the other functions above

File: src/DataStructures/Tensor/EagerMath/DotProduct.hpp [hunk 4]
```diff
@@ -75,6 +75,50 @@ Scalar<DataType> dot_product(
...
+ *
+ * \details
+ * Returns \f$v_a T^a_{b\ldots}\f$, where \f$v_a\f$ is `vector_left`, and
+ * \f$T^a_{b\ldots}\f$ is `tensor_right`.
+ */
+template <typename DataType, typename Index, typename Symm,
+          typename... RemainingIndices,
+          typename ResultTensor = Tensor<DataType, tmpl::pop_front<Symm>,
+                                         index_list<RemainingIndices...>>>
+void dot_product(
```
Comment (kidder) [inline src/DataStructures/Tensor/EagerMath/DotProduct.hpp]: rename to `normal_dot_flux` (we are assuming this is what you want it for...) and move to an appropriate location,
and then it is always a one-form contracted into a tensor

File: tests/Unit/DataStructures/Tensor/EagerMath/Test_DotProduct.cpp [hunk 5]
```diff
@@ -151,6 +151,10 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.EagerMath.DotProduct",
     CHECK_ITERABLE_APPROX(
         get(dot_product(three_d_covector_a, three_d_covector_b, inv_g)),
         DataVector(npts, 486.0));
+    const auto covector_dot_tensor = dot_product(three_d_covector_a, inv_g);
```
Comment (kidder) [inline tests/Unit/DataStructures/Tensor/EagerMath/Test_DotProduct.cpp]: more tests, more dimensions, more tensors (at least to rank 4) and some non-trivial symmetries... (e.g. Ijaa)

**Review by kidder — CHANGES_REQUESTED** [review]
group review with @moxcodes @wthrowe @nilsdeppe @markscheel

**Comment by nilsvu** [comment]
@kidder Made the requested changes. Since all lines of code in this PR changed, I rewrote the commits. Please check the new version :)

File: tests/Unit/Domain/Test_NormalDotFlux.cpp [hunk 6]
```diff
@@ -0,0 +1,346 @@
+// Distributed under the MIT License.
```
Comment (nilsvu) [inline tests/Unit/Domain/Test_NormalDotFlux.cpp]: You can see the correct diff for this file when stepping through the commits one-by-one

**Comment by nilsvu** [comment]
Please check if the changes to this PR resolve the review comments. Travis reports that GCC is complaining about something that I'll fix asap.

@kidder for a group review, could you specify who will follow up on fixups for the PR, i.e. name one of the reviewers "main" or something?

File: tests/Unit/Domain/Test_NormalDotFlux.cpp [hunk 7]
```diff
@@ -145,10 +171,176 @@ void check() {
...
+    check_normal_dot_flux(tnsr::i<DataVector, 1>{{{one}}},
+                          tnsr::I<DataVector, 1>{{{two}}},
+                          Scalar<DataVector>{two});
+    check_normal_dot_flux(tnsr::i<DataVector, 2>{{{one, two}}},
+                          tnsr::I<DataVector, 2>{{{three, four}}},
+                          Scalar<DataVector>{eleven});
+    check_normal_dot_flux(tnsr::i<DataVector, 3>{{{one, two, three}}},
+                          tnsr::I<DataVector, 3>{{{-four, -two, three}}},
+                          Scalar<DataVector>{one});
+    check_normal_dot_flux(tnsr::i<DataVector, 3>{{{one, two, three}}},
```
Comment (kidder) [inline tests/Unit/Domain/Test_NormalDotFlux.cpp]: isn't this a repeat?

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
@nilsdeppe do you have an opinion on where `NormalDotFlux` was moved to?

**Comment by nilsdeppe** [comment]
I'd probably go with `NumericalAlgorithms/DG/NormalDotFluxes.hpp`. Seems the most relevant place at the moment :man_shrugging:

File: tests/Unit/Domain/Test_NormalDotFlux.cpp (hunk 7 above)
Comment (nilsvu) [reply tests/Unit/Domain/Test_NormalDotFlux.cpp]: Yep, will remove.

Do you think I should actually keep these specific tests around or is the Pypp Test enough? I just computed these numbers in Python as well..

**Comment by nilsvu** [comment]
Moved to `NumericalAlgorithms/DG/NormalDotFluxes.hpp` as suggested and deleted the one duplicate test. Please take another look @kidder

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@nilsdeppe please look at

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsvu** [comment]
This is ready to merge. Thanks for reviewing!

**Pull request merged**